src/Adapter/Product/Image/CommandHandler/SetProductImagesForAllShopHandler.php

```diff
@@ -0,0 +1,164 @@
+            $shopsToRemoveImageFrom = $this->getShopsToRemoveImageFrom($shopIdsAssociatedToProduct, $shopsToAddImageTo, $image);
+            $image->associateTo($shopsToAddImageTo, $command->getProductId()->getValue());
+            if (!empty($shopsToRemoveImageFrom)) {
+                $this->productImageMultiShopRepository->deleteFromShops(
+                    $image,
+                    array_map(
+                        static function (int $shopId): ShopId {
+                            return new ShopId($shopId);
+                        },
+                        $shopsToRemoveImageFrom
+                    )
+                );
+            }
+        }
+    }
+
+    /**
+     * @param ProductId $productId
+     *
+     * @return int[]
+     */
+    private function getShopIdsAssociatedToProduct(ProductId $productId): array
+    {
+        $shopIdsAssociatedToProduct = array_map(
+            static function (ShopId $shopId): int {
+                return $shopId->getValue();
+            },
+            $this->productMultiShopRepository->getAssociatedShopIds($productId)
+        );
+
+        return $shopIdsAssociatedToProduct;
+    }
+
+    /**
+     * @param ProductImageSetting[] $productImageSettings
+     * @param int $imageId
+     *
+     * @return ProductImageSetting|null
+     */
+    private function getProductImageSettingByImage(array $productImageSettings, int $imageId): ?ProductImageSetting
+    {
+        $productImageSettingsFiltered = array_filter(
+            $productImageSettings,
+            function (ProductImageSetting $productImageSetting) use ($imageId): bool {
+                return $productImageSetting->getImageId()->getValue() === $imageId;
+            }
+        );
+
+        return reset($productImageSettingsFiltered) ?: null;
+    }
+
+    /**
+     * @param ProductImageSetting[] $productImageSettings
+     * @param int $imageId
+     *
+     * @return array
+     */
+    private function extractShopsToAddImageTo(array $productImageSettings, int $imageId): array
+    {
+        $productImageSetting = $this->getProductImageSettingByImage($productImageSettings, $imageId);
+        $shopsToAddImageTo = [];
+        if ($productImageSetting) {
+            $shopsToAddImageTo = array_map(
+                static function (ShopId $shopId): int {
+                    return $shopId->getValue();
+                },
+                $productImageSetting->getShopIds()
+            );
+        }
+
+        return $shopsToAddImageTo;
+    }
+
+    /**
+     * @param array $shopIdsAssociatedToProduct
+     * @param array $shopsToAddImageTo
+     * @param Image $image
+     *
+     * @return array
+     */
+    private function getShopsToRemoveImageFrom(array $shopIdsAssociatedToProduct, array $shopsToAddImageTo, Image $image): array
+    {
+        $shopsToRemoveImageFrom = array_diff($shopIdsAssociatedToProduct, $shopsToAddImageTo);
+        $shopIdsAssociatedToImage = array_map(
```

Comment:
I feel like we are using this way too often and its probably time to invent some reusable option for Ids VO to int convertion :thinking: (in a dedicated PR). I think you could at least move those array_maps into reusable private method at least for this handler to make code more readable? like:
```
/**
 * @param ShopId[] $shopIds
 */
private function shopIdsToInt(array $shopIds): array {...}

Reply:
I agree we should fine a way to do this conversion. I am going to make the private method